### PR TITLE
Implement compaction filter for storage

### LINKED ISSUE #286: Implement compaction filter for storage
* Data should be removed after related schema has been deleted.
 * TTL and max-version should be supported.

### PR DESCRIPTION
**Pull request by dangleptr**
Close #286

### CODE REVIEW

**Comment by dangleptr** [comment]
Jenkins go

File: src/storage/CompactionFilter.h [hunk 1]
```diff
@@ -8,40 +8,100 @@
...
     }
 
-    const char* Name() const override {
-        return "NebulaCompactionFilter";
+    bool indexValid(const folly::StringPiece&) const {
+        return true;
     }
+
+private:
+    mutable std::string lastKeyWithNoVerison_;
```
Comment (dutor) [inline src/storage/CompactionFilter.h]: Is the filter called from a single thread? If not, `lastKeyWithNoVersion` shoud be protected.

File: src/storage/CompactionFilter.h [hunk 2]
```diff
@@ -8,40 +8,100 @@
...
+    }
+
+    bool ttlValid(GraphSpaceID, const folly::StringPiece&) const {
+        return true;
+    }
+
+    bool filterVersions(const folly::StringPiece& key) const {
+        folly::StringPiece keyWithNoVersion = NebulaKeyUtils::keyWithNoVersion(key);
+        if (keyWithNoVersion == lastKeyWithNoVerison_) {
+            // TODO(heng): we could support max-versions configuration in schema if needed.
```
Comment (zlcook) [inline src/storage/CompactionFilter.h]: `max-versions` configuration means that we can save multiple keys with different versions, Right ?

File: src/storage/CompactionFilter.h (hunk 1 above)
Comment (dangleptr) [reply src/storage/CompactionFilter.h]: Yes,  sub-compaction is closed by default.  Of course, we could check the config again in code.

File: src/storage/CompactionFilter.h (hunk 2 above)
Comment (dangleptr) [reply src/storage/CompactionFilter.h]: Yes,  in some cases, we need to track the history versions of a vertex or edge.  The feature will be implemented in the future.   We call it time-series graph db.  HAHA

File: src/kvstore/CompactionFilter.h [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+    }
+
+    std::unique_ptr<rocksdb::CompactionFilter>
+    CreateCompactionFilter(const rocksdb::CompactionFilter::Context& context) override {
+        auto now = time::TimeUtils::nowInSeconds();
+        if (context.is_full_compaction) {
+            LOG(INFO) << "Do full compaction!";
+            lastRunCustomFilterTimeSec_ = now;
+            return std::unique_ptr<rocksdb::CompactionFilter>(
+                            new KVCompactionFilter(spaceId_, createKVFilter()));
```
Comment (critical27) [inline src/kvstore/CompactionFilter.h]: Do we have to consider default full compaction without filter?

Comment (dangleptr) [reply src/kvstore/CompactionFilter.h]: One important purpose for full compaction is to remove useless data.  So we should use custom filter.

File: src/storage/CompactionFilter.h [hunk 4]
```diff
@@ -8,40 +8,100 @@
...
+        } else {
+            VLOG(3) << "Skip the system key inside, key " << key;
+        }
+        return false;
+    }
+
+    bool schemaValid(GraphSpaceID spaceId, const folly::StringPiece& key) const {
+        if (NebulaKeyUtils::isVertex(key)) {
+            auto tagId = NebulaKeyUtils::getTagId(key);
+            if (schemaMan_->getNewestTagSchemaVer(spaceId, tagId) < 0) {
```
Comment (steppenwolfyuetong) [inline src/storage/CompactionFilter.h]: When there are multiple versions of the data, do we only save the latest data? Are the rest deleted in the compaction?

File: src/kvstore/CompactionFilter.h [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+        : spaceId_(spaceId)
+        , kvFilter_(std::move(kvFilter)) {
+    }
+
+    bool Filter(int,
+                const rocksdb::Slice& key,
+                const rocksdb::Slice& val,
+                std::string*,
+                bool*) const override {
+        return kvFilter_->filter(spaceId_,
```
Comment (steppenwolfyuetong) [inline src/kvstore/CompactionFilter.h]: Could remove some of the parameters that are not used?

Comment (dangleptr) [reply src/kvstore/CompactionFilter.h]: It is rocksdb's interface.

File: src/storage/CompactionFilter.h (hunk 4 above)
Comment (dangleptr) [reply src/storage/CompactionFilter.h]: yes

File: src/storage/CompactionFilter.h [hunk 6]
```diff
@@ -8,40 +8,100 @@
...
+                VLOG(3) << "Space " << spaceId << ", EdgeType " << edgeType << " invalid";
+                return false;
+            }
+        }
+        return true;
+    }
+
+    bool ttlValid(GraphSpaceID, const folly::StringPiece&) const {
+        return true;
+    }
```
[the remark this answers is not in the document]
Comment (dangleptr) [reply src/storage/CompactionFilter.h]: Good point.

File: src/storage/CompactionFilter.h (hunk 2 above)
Comment (whitewum) [reply src/storage/CompactionFilter.h]: awesome

**Comment by dangleptr** [comment]
> Everything looks fine to me. Maybe we can try to use FilterV2 to reduce some IO potentially later.

Good point.  Let's check the filterV2 usage.

**Pull request merged**